source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -993,6 +1062,18 @@ Network::FilterStatus Filter::onNewConnection() {
       std::make_shared<StreamInfo::StreamIdProviderImpl>(config_->randomGenerator().uuid()));
 
   ASSERT(upstream_ == nullptr);
+
+  // Check if we should delay upstream connection establishment.
+  using Trigger = envoy::extensions::filters::network::tcp_proxy::v3::UpstreamConnectTrigger;
+
+  if (connect_mode_ != Trigger::IMMEDIATE && config_->upstreamConnectTrigger().has_value()) {
+    // Delayed connection establishment based on mode.
+    ENVOY_CONN_LOG(debug, "Delaying upstream connection establishment based on configured mode",
+                   read_callbacks_->connection());
+    return Network::FilterStatus::StopIteration;
+  }
+
+  // Immediate connection establishment (default behavior).
```

Comment:
I think you cannot simply return stop iteration here because the ON_DOWNSTREAM_DATA require to receive the data from downstream?